pandas/core/indexes/base.py

```diff
@@ -516,6 +515,12 @@ def _simple_new(cls, values, name=None, dtype=None, **kwargs):
 
         result = object.__new__(cls)
         result._data = values
+        # _index_data is a (temporary?) fix to ensure that the direct data
+        # manipulation we do in `_libs/reduction.pyx` continues to work.
+        # We need access to the actual ndarray, since we're messing with
+        # data buffers and strides. We don't re-use `_ndarray_values`, since
+        # we actually set this value too.
+        result._index_data = values
```

Comment:
Would this be clearer as a property?

Reply:
Probably, but I suspect that if we're going to these lengths in `_libs/reduction.pyx`, the overhead of a property access would be prohibitive.

Reply:
It's mostly over my head, but I can try :)

The docstring for Reducer states

```
    Performs generic reduction operation on a C or Fortran-contiguous ndarray
    while avoiding ndarray construction overhead
```

IIUC, rather than create new ndarrays, we

1. grab a reference to the memory location of the start of an ndarray,
2. operate on that chunk of memory
3. "advance" by setting the memory at that address to the values of the next chunk of memory.

But the usual caveat of "I don't know C" applies here.

Reply:
The reason for this is that it is *much* faster to just user pointer arithmetic on a numpy array of values. Since we wrap these in Index, we need a way to recreate the index (so the inside of the reduction function can see this as an Index), but w/o the overhead of actually doing it. This does make a substantial perf difference as you don't normally *do* much with the index inside the function (usually you only do stuff with the values), so you are needing to recreate the Index each time and then just throwing it away.

This certainly could be refactored I think.